csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -441,295 +394,497 @@ async Task StartedAsync(IServiceObserverPrx observer, string[] services)
+                            return true;
+                        }
+                        else
+                        {
+                            return false;
                         }
-                    }
-
-                    if (info.Communicator != null)
-                    {
-                        DestroyServiceCommunicator(info.Name, info.Communicator);
                     }
                 }
+            }
+
+            internal string Name { get; }
+
+            private readonly string[] _args;
+
+            private readonly object _mutex = new(); // protects _startTask and _stopTask
+
+            private readonly IService _service;
 
-                if (_sharedCommunicator != null)
+            // When _startTask is not null, the service is starting, started (_startTask completed successfully) or
+            // stopped (_startTask is failed or canceled).
+            private Lazy<Task>? _startTask;
+
+            // When _stopTask is not null, the service is stopping, stopped or started (_stopTask is failed or
+            // canceled).
+            private Lazy<Task>? _stopTask;
+
+            internal ServiceInfo(string name, IService service, Communicator? communicator, string[] args)
+            {
+                _args = args;
+                Communicator = communicator;
+                Name = name;
+                _service = service;
+                _startTask = new(Task.CompletedTask); // ServiceInfo is created after the initial successful start
+            }
+
+            internal async Task StartServiceAsync(Communicator? sharedCommunicator, CancellationToken cancel)
+            {
+                // We use lazy tasks to avoid calling Service.StartAsync and Service.StopAsync with _mutex locked
+                // because StartAsync/StopAsync can potentially acquire synchronously other mutexes and create an
+                // out-of-order lock acquisition deadlocks.
+
+                Lazy<Task>? performTask = null;
+
+                do
                 {
-                    RemoveAdminFacets("IceBox.SharedCommunicator.");
+                    // We may need to wait until the current stop task completes.
+                    Lazy<Task>? waitTask = null;
 
-                    try
+                    lock (_mutex)
                     {
-                        _sharedCommunicator.Dispose();
+                        if (_stopTask is Lazy<Task> stopTask)
+                        {
+                            Debug.Assert(_startTask == null);
+                            if (!stopTask.IsValueCreated)
+                            {
+                                waitTask = stopTask;
+                            }
+                            else if (stopTask.Value.IsFaulted || stopTask.Value.IsCanceled)
+                            {
+                                // Already started, no need to do anything.
```

Comment:
It's not obvious why this means the service is already started.

Reply:
It's from the semantics of StartAsync and StopAsync on IService (see remarks), e.g.:
```
        /// <remarks>If StopAsync throws an exception, the service manager consider the service remains in the started
        /// started.</remarks>
```

Reply:
Changed this code quite a bit in latest commit.